### DISCUSSION TITLE
How to convert fabric filter values of version 1.x to 4.x

### DISCUSSION THREAD

**Discussion by shabeernotion** [fabricjs/fabric.js#6760] [Q&A]
I am using fabric Js version 1.6.4 on canvas, but I am using fabric Js version 4.x while preview the large size.
Filters such as Blur, Brightness, Noise, Hue, Saturation and Contrast values are range from 1 to 100 in fabric 1.x. How to convert this values to 4.x filter values which is ranging between 0 and 1 or -1 and 1. However I have managed to convert few filters by referring [this][1] .

For example:

    Brightness range 1.x = 70
    (Conversion to 4.x) = (Brightness range 1.x) / 255

    Contrast range 1.x = 50
    (Conversion to 4.x) = (Contrast range 1.x) / 255

However I am not able to convert Noise using the above calculations, because Noise filter is ranging between 100 to 1000 in 4.x.
I have tried doing like the following.

            {   name: "Noise",
    			obj: function (range) {
    				range = (range / 255) * 100;
    				return new fabric.Image.filters.Noise({ 
    					noise: parseFloat(range)
    				}); 
    			}
    		}

But it is not working as expected, However following method gives the closest result, but it is not officially documented, I have assumed the multiplier of 200.

           {    name: "Noise", 
    			obj: function (range) {
    			range = (range / 255) * 200; /* Assumed 200 multiplier because there were no doc how to convert 
    				return new fabric.Image.filters.Noise({ 
    					noise: parseFloat(range)
    				}); 
    			}
    		}

 1. Why do we have to divide brightness by 255 while actually we have to divide by 100 to get the range between 0 to 1?
 2. How to convert Noise and Blur(Not working as expected if divided old value by 255, dividing 100 give closest result) values of 1.x version to match 4.x version?

Please refer if any other official doc is out there for all filter conversions.

  [1]: http://fabricjs.com/fabric-filters

**Comment by asturur** [comment]
There are suggestions on fabricjs.com on how to deal with it.
http://fabricjs.com/fabric-filters

I would suggest you for your own peace of mind to just move to the same version everywhere.

**Reply by shabeernotion** [reply]
@asturur I am not able to use that old version using NPM packages. It's deprecated on NPM. The link does not have full explanation on how to deal with all filters. Noise filter and Blur getting problem.

**Reply by asturur** [reply]
I was saying move everything on the latest.
in the artcle i linked:

```
fabric.Image.filters.BaseFilter.fromObject = function(object, callback) {
  switch (object.type) {
    case 'Brightness':
      object.brightness = object.brightness / 255;
    case 'Contrast':
      object.contrast = object.contrast / 255;
    break;
    case 'Mask':
      object.type = 'BlendImage';
      object.image = object.mask;
    break;
    case 'Blend':
      if (!object.image) {
        object.type = 'BlendColor';
      } else {
        // conversion harder
      }
    break;
    case 'Multiply':
      object.type = 'BlendColor';
      object.mode = 'multiply';
    break;
    // may not give exact same result
    case 'RemoveWhite':
      object.type = 'RemoveColor';
      object.color = '#FFFFFF';
      object.distance = object.distance / 255;
    break;
    case 'Saturate':
      object.saturation = object.saturate / 100;
    break;
    case 'Tint':
      object.type = 'BlendColor';
      object.saturation = object.saturate / 100;
    break;
    // Color matrix in 1.x where made for imageData and the column of the costants
    // was in the -255 to 255 range, now is in the -1 to 1
    case 'ColorMatrix':
      object.matrix[4] = object.matrix[4] / 255;
      object.matrix[9] = object.matrix[9] / 255;
      object.matrix[14] = object.matrix[14] / 255;
      object.matrix[19] = object.matrix[19] / 255;
    break;
  }
  var filter = new fabric.Image.filters[object.type](object);
  callback && callback(filter);
  return filter;
};
```
What filter is missing in this example? we can add what is missing

**Reply by asturur** [reply]
urgh sub threading. This wasn't my intention, please reply on the main thread

**Comment by shabeernotion** [comment]
We are planning to move to the latest version before that we need check, if we are able to convert the 1.x values to 4.x.
Blur and Noise filters are missing on the above document.

Sorry for sub threading.

**Comment by asturur** [comment]
Noise does not have a 1:1 match being a random effect, sadly, but for blur we can fix it.
i ll add those missing ones.
Does the solution works for you anyway provided the missing filters?

**Comment by shabeernotion** [comment]
Okay, Please add solution for blur. All filters are working fine, but saturation/100 not works for me, but saturation/255 gives closest results for me.
